src/kixi/lapita.clj

```diff
@@ -20,38 +20,57 @@
-   If a schema is provided, the data is coerce to the right type."
-  ([filename]
-   (let [file (io/file filename)]
-     (when (.exists (io/as-file file))
-       (let [parsed-csv (with-open [in-file (io/reader file)]
-                          (doall (data-csv/read-csv in-file)))
-             parsed-data (rest parsed-csv)
-             headers (map format-key (first parsed-csv))]
-         (ds/dataset (map #(zipmap headers %) parsed-data))))))
-  ([filename schema]
-   (-> (let [file (io/file filename)]
-         (when (.exists (io/as-file file))
-           (let [parsed-csv (with-open [in-file (io/reader file)]
-                              (doall (data-csv/read-csv in-file)))
-                 parsed-data (rest parsed-csv)
-                 headers (map format-key (first parsed-csv))]
-             {:column-names headers
-              :columns (vec parsed-data)})))
-       (sc/apply-schema-coercion schema)
-       (as-> {:keys [column-names columns]} (ds/dataset column-names columns)))))
-
 (defn write-csv!
-  "Write a dataset to a csv file"
-  [ds f]
-  (let [rows-as-maps (ds/row-maps ds)
+  "Write data (dataset or collection of maps) to a csv file"
+  [data f]
+  (let [rows-as-maps (if (ds/dataset? data) (ds/row-maps data) data)
         colnames (mapv name (keys (first rows-as-maps)))
         rows (mapv #(vec (vals %)) rows-as-maps)]
     (with-open [out-file (io/writer f)]
       (data-csv/write-csv out-file (into [colnames] rows)))))
 
+(defn csv-to-maps-coll
+  "Takes in a filename, return a map containing the column header
+   and the column content in a vector of vectors."
+  [filename]
+  (let [file (io/file filename)]
+    (when (.exists (io/as-file file))
+      (let [parsed-csv (with-open [in-file (io/reader file)]
+                         (doall (data-csv/read-csv in-file)))
+            parsed-data (rest parsed-csv)
+            headers (map format-key (first parsed-csv))]
+        (map #(zipmap headers %) parsed-data)))))
+
+(defn csv-to-dataset
+  "Loads csv file, returns a core.matrix dataset.
+   If a schema is provided alone, the data is coerced to the right type
+   unless it leads to a coercion error.
+   If a schema is provided with a map of options, then you can return only
+   data consistent with the schema and either print or save to file the data
+   that fails data coercion.
+   Example of options {:print-errors true :write-errors false :remove-errors true}"
+  ([filename]
+   (ds/dataset (csv-to-maps-coll filename)))
+  ;; Returns a mix of coerced data and data inconsistent w/ schema
+  ([filename schema]
+   (let [all-data (csv-to-maps-coll filename)
+         coerced-data (sc/coerce-data-from-schema all-data schema)]
+     (ds/dataset coerced-data)))
+  ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
+  ;; write inconsistent data to file return only coerced data.
+  ([filename schema {:keys [print-errors write-errors remove-errors]}]
```

Comment:
When you have an arity with optional additional 'options' it's usual that the version without (two-arity in this case) calls the version with (three-arity in this case), just with an empty options map:

```
(defn foo
  ([x y opts]
    ...)
  ([x y]
    (foo x y {}))
```